mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -1518,24 +1511,236 @@ def _log_inputs_impl(
+            order_by: List of order_by clauses.
+            page_token: Token specifying the next page of results. It should be obtained from
+                a ``search_traces`` call.
+
+        Returns:
+            A tuple of a list of :py:class:`TraceInfo <mlflow.entities.TraceInfo>` objects that
+            satisfy the search expressions and a pagination token for the next page of results.
+        """
+        self._validate_max_results_param(max_results)
+
+        with self.ManagedSessionMaker() as session:
+            cases_orderby, parsed_orderby, sorting_joins = _get_orderby_clauses_for_search_traces(
+                order_by or [], session
+            )
+            stmt = select(SqlTraceInfo, *cases_orderby)
+
+            attribute_filters, non_attribute_filters = _get_filter_clauses_for_search_traces(
+                filter_string, session, self._get_dialect()
+            )
+            for non_attr_filter in non_attribute_filters:
+                stmt = stmt.join(non_attr_filter)
+
+            # using an outer join is necessary here because we want to be able to sort
+            # on a column (tag, metric or param) without removing the lines that
+            # do not have a value for this column (which is what inner join would do)
+            for j in sorting_joins:
+                stmt = stmt.outerjoin(j)
+
+            offset = SearchTraceUtils.parse_start_offset_from_page_token(page_token)
+            stmt = (
+                stmt.distinct()
+                .filter(
+                    SqlTraceInfo.experiment_id.in_(experiment_ids),
+                    *attribute_filters,
+                )
+                .order_by(*parsed_orderby)
+                .offset(offset)
+                .limit(max_results)
+            )
+            queried_traces = session.execute(stmt).scalars(SqlTraceInfo).all()
+            trace_infos = [t.to_mlflow_entity() for t in queried_traces]
+
+            # Compute next search token
+            if max_results == len(trace_infos):
+                final_offset = offset + max_results
+                next_token = SearchTraceUtils.create_page_token(final_offset)
+            else:
+                next_token = None
+
+            return trace_infos, next_token
+
+    def _validate_max_results_param(self, max_results: int, allow_null=False):
+        if (not allow_null and max_results is None) or max_results < 1:
+            raise MlflowException(
+                "Invalid value for request parameter max_results. It must be "
+                f"a positive integer, but got {max_results}",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        if max_results > SEARCH_MAX_RESULTS_THRESHOLD:
+            raise MlflowException(
+                "Invalid value for request parameter max_results. It must be at "
+                f"most {SEARCH_MAX_RESULTS_THRESHOLD}, but got {max_results}",
+                INVALID_PARAMETER_VALUE,
+            )
+
+    def delete_traces(
```

Comment:
When we effectively hard-delete the trace info from the tables, what is the mechanism for artifact store cleanup of the associated trace data? 
If we call mlflow.gc() with the changes that will need to be implemented to handle trace data cleanup, do we delete directories within experiments that do not have a corresponding request_id entry? 
This is a bit of a departure from how delete_runs and delete_experiment functions (with their setting of a lifecycle stage which is evaluated for hard-deletion and artifact store cleanup of serialized artifacts when calling mlflow.gc). 
Can we align this API with the mechanism for artifact store trace data cleanup?

Reply:
[note] As discussed offline, we will follow-up for this soft-vs-hard deletion topic.

Reply:
Let's go the hard-deletion route. This matches with the source implementation on the Databricks side :)